Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in PAL

cdumez opened the pull request:
#### 76e39111b6161c3afc4f0cbbd1d722e11bb1b919
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in PAL
<a href="https://bugs.webkit.org/show_bug.cgi?id=285236">https://bugs.webkit.org/show_bug.cgi?id=285236</a>

Reviewed by Darin Adler.

* Source/WebCore/PAL/pal/Gunzip.h:
* Source/WebCore/PAL/pal/avfoundation/OutputContext.mm:
(PAL::OutputContext::deviceName):
* Source/WebCore/PAL/pal/cocoa/Gunzip.cpp:
(PAL::gunzip):
* Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp:
* Source/WebCore/PAL/pal/text/EncodingTables.cpp:
(PAL::jis0208):
(PAL::jis0212):
(PAL::big5):
(PAL::eucKR):
(PAL::gb18030):
* Source/WebCore/PAL/pal/text/TextCodecASCIIFastPath.h:
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::shiftJISEncode):
* Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
(PAL::TextCodecICU::registerEncodingNames):
(PAL::TextCodecICU::decode):
(PAL::TextCodecICU::encode const):
* Source/WebCore/PAL/pal/text/TextCodecSingleByte.cpp:
(PAL::tableForEncoding):
* Source/WebCore/PAL/pal/text/TextCodecUTF16.cpp:
(PAL::TextCodecUTF16::decode):
(PAL::TextCodecUTF16::encode const):
* Source/WebCore/PAL/pal/text/TextCodecUTF8.cpp:
(PAL::nonASCIISequenceLength):
(PAL::TextCodecUTF8::decode):
(PAL::TextCodecUTF8::encodeUTF8):
* Source/WebCore/PAL/pal/text/TextCodecUserDefined.cpp:
(PAL::TextCodecUserDefined::encode const):

Canonical link: <a href="https://commits.webkit.org/288345@main">https://commits.webkit.org/288345@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f6bd2e128c0c31e28ce13af3bcf303c1b869a61

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82673 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2324 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36828 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87793 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33731 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84767 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2396 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10235 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64420 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22191 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85730 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1747 "Found 1 new test failure: fast/loader/redirect-to-invalid-url-using-meta-refresh-calls-policy-delegate.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75203 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44695 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1642 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29389 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32765 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72918 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30048 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89159 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9966 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7218 "Found 2 new test failures: http/tests/websocket/tests/hybi/websocket-constructor-protocols.html imported/w3c/web-platform-tests/mediacapture-record/MediaRecorder-peerconnection.https.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72835 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10194 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71012 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72049 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16200 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1355 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12812 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9919 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15440 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9793 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13258 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11562 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/PAL/pal/avfoundation/OutputContext.mm:
```diff
@@ -77,16 +75,11 @@
+        builder.append(device.name());
     }
-
     return builder.toString();
```
@weinig added something to `StringConcatenate.h` to do this with a lambda that I think is nicer. I can't remember the syntax offhand.

cdumez replied:
Probably `interleave()`

darinadler commented on Source/WebCore/PAL/pal/text/EncodingTables.cpp:
```diff
@@ -1073,16 +1072,19 @@ const std::array<std::pair<uint16_t, UChar>, 7724>& jis0208()
+                UChar* outputEnd = std::to_address(outputSpan.end());
+                const char* input = byteCast<char>(icuInput.data());
+                const char* inputEnd = byteCast<char>(std::to_address(icuInput.end()));
+                ucnv_toUnicode(icuConverter.get(), &output, outputEnd, &input, inputEnd, nullptr, true, &error);
```
I think it might be safer to have a single cover for `ucnv_toUnicode` called in all these places so the length handling and pointer stuff can be done right just once.

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
```diff
@@ -677,7 +675,8 @@ static Vector<uint8_t> shiftJISEncode(StringView string, Function<void(char32_t,
         }
 
-        ASSERT(range.first + 3 >= range.second);
+        ASSERT(range.second - range.first <= 3);
```
Do we also need to assert that `range.second >= range.first` to avoid underflow? Maybe not. I'm unclear on the types involved.

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
```diff
@@ -232,22 +229,22 @@ String TextCodecICU::decode(std::span<const uint8_t> bytes, bool flush, bool sto
     do {
-        size_t ucharsDecoded = decodeToBuffer(buffer, bufferLimit, source, sourceLimit, offsets, flush, err);
-        result.append(std::span { buffer, ucharsDecoded });
+        size_t ucharsDecoded = decodeToBuffer(buffer.data(), bufferLimit, source, sourceLimit, offsets, flush, err);
```
`decodeToBuffer` should take `span` arguments.

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecLatin1.cpp:
```diff
@@ -235,18 +235,16 @@ static Vector<uint8_t> encodeComplexWindowsLatin1(StringView string, Unencodable
     {
         Vector<uint8_t> result(string.length());
-        auto* bytes = result.data();
+        size_t resultIndex = 0;
```
Not sure this needs a two word name!

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecSingleByte.cpp:
```diff
@@ -180,22 +180,21 @@ template<const SingleByteDecodeTable& decodeTable> SingleByteEncodeTable tableFo
     // Allocate this at runtime because building it at compile time would make the binary much larger and this is often not used.
     static constexpr auto size = std::size(decodeTable) - std::count(std::begin(decodeTable), std::end(decodeTable), replacementCharacter);
-    static const SingleByteEncodeTableEntry* entries;
+    static LazyNeverDestroyed<std::array<SingleByteEncodeTableEntry, size>> entries;
```
Does this make the binary larger the way the comment says we should not?

cdumez replied:
I think you're right, I'll use a raw pointer.

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecSingleByte.cpp:
```diff
@@ -180,22 +180,21 @@ template<const SingleByteDecodeTable& decodeTable> SingleByteEncodeTable tableFo
         }
         ASSERT(j == size);
-        auto collection = WTF::makeIteratorRange(&mutableEntries[0], &mutableEntries[size]);
+        auto collection = WTF::makeIteratorRange(&entries.get()[0], &entries.get()[size]);
```
Why do pointers need to be involved? Isn't a `span` a collection we can pass to the functions?

cdumez replied:
`WTF::makeIteratorRange()` seems to take 2 iterators at the moment, no container. But I can fix that.

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecSingleByte.cpp:
```diff
@@ -180,22 +180,21 @@ template<const SingleByteDecodeTable& decodeTable> SingleByteEncodeTable tableFo
-        entries = mutableEntries;
     });
-    return WTF::makeIteratorRange(&entries[0], &entries[size]);
+    return WTF::makeIteratorRange(const_cast<const SingleByteEncodeTableEntry*>(entries.get().data()), const_cast<const SingleByteEncodeTableEntry*>(std::to_address(entries.get().end())));
```
Can this be changed to return a `span` instead of an "iterator range"?

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecUTF16.cpp:
```diff
@@ -149,26 +149,26 @@ String TextCodecUTF16::decode(std::span<const uint8_t> bytes, bool flush, bool,
 {
     Vector<uint8_t> result(WTF::checkedProduct<size_t>(string.length(), 2));
-    auto* bytes = result.data();
+    size_t resultIndex = 0;
```
I'd prefer a shorter one word name.

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecUserDefined.cpp:
```diff
@@ -75,18 +75,16 @@ static Vector<uint8_t> encodeComplexUserDefined(StringView string, UnencodableHa
     {
         Vector<uint8_t> result(string.length());
-        auto* bytes = result.data();
+        size_t resultIndex = 0;
```
I'd prefer a shorter one word name.

darinadler commented on Source/WebCore/PAL/pal/avfoundation/OutputContext.mm:
```diff
@@ -77,16 +76,9 @@
-            builder.append(" + "_s);
-    }
-
+    builder.append(interleave(outputDevices(), [](auto& device) {
```
I don't think we need to explicitly define a `StringBuilder`, just `makeString`.

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in PAL
https://bugs.webkit.org/show_bug.cgi?id=285236

Reviewed by Darin Adler.

* Source/WebCore/PAL/pal/Gunzip.h:
* Source/WebCore/PAL/pal/avfoundation/OutputContext.mm:
(PAL::OutputContext::deviceName):
* Source/WebCore/PAL/pal/cocoa/Gunzip.cpp:
(PAL::gunzip):
* Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp:
* Source/WebCore/PAL/pal/text/EncodingTables.cpp:
(PAL::jis0208):
(PAL::jis0212):
(PAL::big5):
(PAL::eucKR):
(PAL::gb18030):
* Source/WebCore/PAL/pal/text/TextCodecASCIIFastPath.h:
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::shiftJISEncode):
* Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
(PAL::TextCodecICU::registerEncodingNames):
(PAL::TextCodecICU::decode):
(PAL::TextCodecICU::encode const):
* Source/WebCore/PAL/pal/text/TextCodecSingleByte.cpp:
(PAL::tableForEncoding):
* Source/WebCore/PAL/pal/text/TextCodecUTF16.cpp:
(PAL::TextCodecUTF16::decode):
(PAL::TextCodecUTF16::encode const):
* Source/WebCore/PAL/pal/text/TextCodecUTF8.cpp:
(PAL::nonASCIISequenceLength):
(PAL::TextCodecUTF8::decode):
(PAL::TextCodecUTF8::encodeUTF8):
* Source/WebCore/PAL/pal/text/TextCodecUserDefined.cpp:
(PAL::TextCodecUserDefined::encode const):

Canonical link: https://commits.webkit.org/288345@main
